share/wake/lib/core/json.wake

```diff
@@ -154,18 +154,155 @@ export def x // y =
+        | rmap JDouble
+    JBoolean a =
+        getJSONNormalizeBoolean fmt a
+        | rmap JBoolean
+    JNull =
+        Pass JNull
+    JObject a =
+        getJSONNormalizeObject fmt a
+        | rmap JObject
+    JArray a =
+        getJSONNormalizeArray fmt a
+        | rmap JArray
+
+# The rules by which `normalizeJSON` will simplify JSON values.
+# If any particular function is unable to operate on every input value (e.g. a
+# `JObject` contains duplicate keys of types which cannot be combined), that
+# rule may return a `Fail` which halts the broader processing.
+export tuple JSONNormalize =
+    String: String => Result String Error
+    Integer: Integer => Result Integer Error
+    Double: Double => Result Double Error
+    Boolean: Boolean => Result Boolean Error
+    Object: List (Pair String JValue) => Result (List (Pair String JValue)) Error
+    Array: List JValue => Result (List JValue) Error
+
+# A basic normalization ruleset which returns inputs unchanged.
+export def normalizeJSONDefault: JSONNormalize =
+    JSONNormalize Pass Pass Pass Pass Pass Pass
+
+# Target the minimum JSON language described by the specifications.
+export def normalizeJSONBasic: JSONNormalize =
+    def nfkcKeys =
+        map (_ | editPairFirst unicodeIdentifier)
+    normalizeJSONDefault
+    | setJSONNormalizeString (Pass _.unicodeCanonical)
+    | setJSONNormalizeDouble filterNonDigitJDouble
+    | setJSONNormalizeObject (deduplicateJObjectKeys _.nfkcKeys)
+
+# Fail on any `Double` values which can't be represented numerically.
+# Both JSON specifications describe their double values as allowing decimal or
+# exponential forms, but don't implement the full IEEE standard.
+def filterNonDigitJDouble (n: Double): Result Double Error = match n.dclass
+    DoubleInfinite =
+        failWithError "For compatibility, JSON doubles should not be infinite."
+    DoubleNaN =
+        failWithError "For compatibility, JSON doubles should not be NaN values."
+    _ =
+        Pass n
+
+# Attempt to simplify any `JObject`s which contain multiple instances of a key.
+# While both specifications explicitly allow such duplicate keys, neither
+# requires implementations to do so or describes the behaviour by which they
+# should be handled; therefore, implementations differ in how they handle
+# duplicate keys -- if they do so at all -- and a generator aiming for
+# compatibility shouldn't output objects with duplicate keys.
+#
+# In particular, this function will pass if all instances of the same key also
+# have the same value, or otherwise will combine the values of `JArray` or
+# `JObject` values if all instances of the key have the same type.  It fails if
+# a key is associated with multiple values of different types or different
+# values of a type which cannot be combined.
+def deduplicateJObjectKeys (dict: List (Pair String JValue)): Result (List (Pair String JValue)) Error =
+    def cmpKeysOnly (Pair k1 _) (Pair k2 _) =
+        k1 <~ k2
```

Comment:
This should probably be `<*` so that you leave the normalization policy to the caller (who already calls `unicodeIdentifier` or something else appropriate).